sdk/eventhub/Azure.Messaging.EventHubs/design/event-processor{T}-proposal.md

```diff
@@ -0,0 +1,520 @@
+    
+    // Required extension points (core)
+    protected abstract Task OnProcessingEventBatchAsync(IEnumerable<EventData> events, TPartition partition, CancellationToken cancellationToken);
+    protected abstract Task OnProcessingErrorAsync(Exception exception, TPartition partition, string operationDescription, CancellationToken cancellationToken);
+    
+    // Required extension points (storage operations)
+    protected abstract Task<IEnumerable<EventProcessorCheckpoint>> ListCheckpointsAsync(CancellationToken cancellationToken);
+    protected abstract Task<IEnumerable<PartitionOwnership>> ListOwnershipAsync(CancellationToken cancellationToken);
+    protected abstract Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> desiredOwnership, CancellationToken cancellationToken);
+    
+    // Infrastructure
+    protected virtual LastEnqueuedEventProperties ReadLastEnqueuedEventProperties(string partitionId);
+}
+
+public class EventProcessorOptions 
+{
+    public EventHubConnectionOptions ConnectionOptions { get; set; }
+    public EventHubsRetryOptions RetryOptions { get; set; }
+    public string Identifier { get; set; }
+    public int? PrefetchCount { get; set; } 
+    public bool TrackLastEnqueuedEventProperties { get; set; }
+    public TimeSpan MaximumWaitTime { get; set; } = TimeSpan.FromSeconds(60);
+    public EventPosition DefaultStartingPosition { get; set; } = EventPosition.Earliest;
+    public TimeSpan LoadBalancingUpdateInterval { get; set; } = TimeSpan.FromSeconds(10);
+    public TimeSpan PartitionOwnershipExpirationInterval { get; set; } = TimeSpan.FromSeconds(30);
+}
+
+public class EventProcessorPartition
+{
+    public string PartitionId { get; set; }
+}
+
+public class EventProcessorCheckpoint
+{
+    public string FullyQualifiedNamespace { get; set; }
+    public string EventHubName { get; set; }
+    public string ConsumerGroup { get; set; }
+    public string PartitionId { get; set; }
+    public long? Offset { get; set; }
+    public long? SequenceNumber { get; set; }
+}
+
+public class PartitionOwnership
+{
+    public string FullyQualifiedNamespace { get; set; }
+    public string EventHubName { get; set; }
+    public string ConsumerGroup { get; set; }
+    public string OwnerIdentifier { get; set; }
+    public string PartitionId { get; set; }
+    public string Version { get; set; }
+    public DateTimeOffset LastModifiedTime { get; set; }
+}
+```
+
+## Namespace organization
+
+### Package: Azure.Messaging.EventHubs
+
+#### `Azure.Messaging.EventHubs.Specialized`
+```csharp
+public class EventProcessor<TPartition> {}
+public class EventProcessorOptions {}
+public class EventProcessorCheckpoint {}
+public class EventProcessorPartition {}
+public class PartitionOwnership {}
+public class PartitionReceiverClient {}
+public class PartitionReceiverClientOptions {}
```

Comment:
Why don't we use the Ownership and Checkpoint already defined in the Shared package?

Reply:
Because they have to move to the core library to be made public.   We can't have public types that exist in two packages without collisions.